Extra chat additions: blinking cursor and new GAMEEVENT_CHAT event type

AKMDM opened the pull request:
These are some extra commits I created, that add a few things to the chat.

- Players cannot name themselves as "server" in online games anymore. This averts potential impersonation of the server and the ambiguous case that if someone tries sending a private message using `sayto server`, whether the server or the player with the same name should be chosen.
- The chat cursor now blinks while the playing is creating a message. If they're typing or passing any input, the cursor remains visible. This always takes into account the width of the '_' character, whether it's visible or not, which does make the implementation somewhat complicated but necessary.
- I touched up and re-added the new EVENT type: GAMEEVENT_CHAT and the ACS function GetChatMessage( int player, int index ). Up to three chat messages from the player (or server) can be stored in memory, and modders can also pass an index between 0 and 2 to retrieve the corresponding chat message, with 0 being the newest and 2 being oldest. This is in case the local machine, whether it's the client or the server, receives more than one chat message from the sender in a single tic, which means the last chat message will replace the older ones.

TorrSamaho commented on src/chat.cpp:
```diff
@@ -692,8 +704,14 @@ void CHAT_Render( void )
+
 	// Insert the cursor string into the message.
-	displayString = displayString.Mid( 0, g_ChatBuffer.GetPosition() ) + cursor + displayString.Mid( g_ChatBuffer.GetPosition() );
+	displayString = messageFront + cursor + messageBack;
```
I may be overlooking something, it shouldn't it be sufficient simply to insert `cursor`  or `" "` depending on whether the cursor should be invisible or not? Then you shouldn't need to change anything else here.

AKMDM replied:
That is how I originally did it, but I found that the problem with doing it that way is that the widths of the graphics used for the '_' and " " characters of the SmallFont are different. So whenever the cursor is not at the end of the message, you'll notice that when the cursor switches between visible and invisible, everything behind the cursor would shift left and right, and this didn't look good. This isn't noticeable when the cursor is at the end of the string.

TorrSamaho replied:
I see. There should be other invisible characters (e.g. tab). Do you know if all have them have same widths? If not, perhaps one of them has the same as `_`.

AKMDM replied:
I don't think there's anything for the tab, since it's a control code. I'm still not sure if it's a good idea to try using another invisible character as a substitute for the cursor when it's invisible. While they might have the same widths when using the standard SmallFont in Doom, what if somebody adds a replacement to the SmallFont and these character pairs end up having different widths?

TorrSamaho commented:
FYI, I added the "Players cannot name themselves as "server" in online games anymore." commit.

TorrSamaho commented:
Can you comment on use cases for the event type GAMEEVENT_CHAT?

AKMDM commented:
- The biggest example for `GAMEEVENT_CHAT` would be manipulating actors or monsters that can understand and react to what players might say in the chat. For example, I had plans to add "fake" innocent bots in Who Dun It that could understand what other players are saying. So if someone said something like "player is the murderer, kill him", the actor would react and target that player. 

- Another example I can think of is for features that lets players _speak_ in a TTS-like voice whatever they typed. Being able to retrieve whatever message the player sent would make this easier to implement. I was also going to add something like this into Who Dun It so that players could speak whatever they typed.

- Chat messages may be used as a way to execute commands within a mod, where using CCMDs might not necessarily work, especially when strings are involved.

TorrSamaho commented:
Thanks for the use cases! This sounds very useful for mods.

TorrSamaho commented on src/chat.cpp:
```diff
@@ -912,13 +980,31 @@ void CHAT_PrintChatString( ULONG ulPlayer, ULONG ulMode, const char *pszString )
+			if ( players[ulPlayer].ChatMessages.Size() >= MAX_STORED_MESSAGES )
+				players[ulPlayer].ChatMessages.Delete( MAX_STORED_MESSAGES - 1 );
+
+			players[ulPlayer].ChatMessages.Insert( 0, message );
```
A `TArray` is not a very efficient data structure for this, because you are copying the entire existing content of the TArray when you add something in front. For this kind of use, a ring buffer would be much more efficient. I think you can use `RingBuffer` from `networkshared.h` instead here. It's designed for keeping the last `N` entries. You may just have to generalize `RingBuffer::clear()`, since it was not created for `DataType`'s that are classes with their own memory management (like `FString`).

AKMDM replied:
I see, thank you for the suggestion. Would changing the data type from `FString` to `char *` work better with the ring buffer?

EDIT: probably not, it might be easier to stick with `FString`.

TorrSamaho replied:
Yeah, sticking with an FString should be the easiest. I think the only thing necessary should be to call the destructor on each stored `FString` in `RingBuffer::clear()`, i.e. only the line
```memset( _data, 0, sizeof( _data ))```
needs adjustment.

EDIT: Looks like you already have taken care of this :).

TorrSamaho commented on src/d_player.h:
```diff
@@ -636,6 +636,9 @@ class player_t
 	bool		bChatting;
 
+	// [AK] The last few chat messages we received from this player.
+	TArray<FString>		ChatMessages;
```
I'm not sure whether the chat messages should be associated with the player directly. It's quite counterintuitive to me that one has to copy chat messages if a player is reborn. I assume you put them here since you didn't want do associate them with the client so that this also works offline?

AKMDM replied:
I want to make sure this could work offline and online (both on the client's and server's ends). Initially, I thought about adding an `FString` array to `chat.cpp` that holds the chat messages for all players + one extra element for the server. If a player left, that element would need to be emptied. However, I was hesitant about including `chat.h` to `p_acs.cpp` so I ended up adding the array directly into `player_t`. This might be counter-intuitive as you said.

TorrSamaho replied:
Putting this into a `RingBuffer` in `chat.cpp` in your latest patch looks good.

TorrSamaho commented on src/p_acs.cpp:
```diff
@@ -7575,7 +7576,6 @@ doplaysound:			if (funcIndex == ACSF_PlayActorSound)
 		case ACSF_GetPlayerScore:
 			{
 				const ULONG ulPlayer = static_cast<ULONG> ( args[0] );
-
```
Looks like this change accidentally slipped in. This code has nothing to do with what the patches are doing.

TorrSamaho replied:
I think this is fixed in your latest patch.

TorrSamaho commented on src/sv_main.cpp:
```diff
@@ -3717,6 +3755,30 @@ void SERVER_SetCurrentFont( const char *pszFont )
+	if ( g_ChatMessages.Size() >= MAX_STORED_MESSAGES )
+		g_ChatMessages.Delete( MAX_STORED_MESSAGES - 1 );
+
+	g_ChatMessages.Insert( 0, message );
```
Here a ring buffer would also be nicer than the array.

AKMDM commented:
I added changes to this pull request, based on the comments from last week. I got rid of the old commit for the `CHATEVENT` and replaced it with the new one which now uses `RingBuffer` to manage the saved chat messages. There's a commit before this one that generalizes the `RingBuffer` class to support data types that are classes and have their own memory management, like you suggested.

The saved chat messages are no longer directly associated with `player_t`, but are instead a global array inside `chat.cpp` that hold a slot for all 64 players + 1 for the server. The elements are reset if a player disconnects (e.x. when a bot leaves in a singleplayer game), and the entire array is reset when a new game starts (e.x. using the "map" CCMD).

I also increased the number of saved messages per player from 3 to 5.

EDIT: Added another commit that deletes the memory in the `_data` array whenever it needs to be cleared, to be prevent garbage collection.

TorrSamaho commented on src/networkshared.h:
```diff
@@ -559,6 +559,9 @@ class RingBuffer
 	void clear ( )
 	{
+		if ( _data != NULL )
+			delete[] _data;
```
You'll also need to call `clear()` in the destructor of `RingBuffer` to free the memory when the instance is destroyed when it falls out of scope.

AKMDM replied:
I added another commit which adds a destructor inside `RingBuffer` to call `clear()`.

TorrSamaho replied:
Thanks!

TorrSamaho commented on src/chat.cpp:
```diff
@@ -781,6 +852,28 @@ ULONG CHAT_GetChatMode( void )
+const char *CHAT_GetChatMessage( ULONG ulPlayer, ULONG ulOffset )
+{
+	FString message = g_SavedChatMessages[ulPlayer].getOldestEntry( ulOffset );
+	return ( message.GetChars( ));
```
I think this may break: You return a pointer to the data stored in the `FString message`, but this `FString` leaves the scope once the function is exited, thus it's destroyed and its memory freed. Why don't you return 
`g_SavedChatMessages[ulPlayer].getOldestEntry( ulOffset ).GetChars( )?`

AKMDM replied:
Sure thing, I'll change that quickly.

TorrSamaho commented:
Thanks for the update! The code looks fine now, but I still have a problem on my end. I'm not sure whether something went wrong during compression and rebasing or if there already was a problem to begin with. Here is my rebased and compressed version:

https://osdn.net/projects/zandronum/scm/hg/zandronum-sandbox-stable/commits/8859c631d9c1a9a06c201b75c99c34d94dd89306

As soon as I join a server, I'm kicked for "Client command flood".

AKMDM commented:
Hi Torr, I tried compiling that version of `zandronum-sandbox-stable` and I also tried porting the same compressed commit into `zandronum-stable` then compiling that, but I can't seem to reproduce the "client command flood" when I try joining a server in both cases. It's kind of strange you're getting this issue, as the changes here don't have the clients send any commands to the server.

Perhaps something else is causing this issue?

EDIT: I added another commit to this pull request. This fixes the saved chat messages from being reset on the client's end every time a level changes (this should only happen when a new game is started or if the client disconnects from the server). I also changed the `RingBuffer` destructor method to only `delete[] _data` instead of calling `clear`, which ended up reallocating the memory and could cause memory leaks.

TorrSamaho commented:
> Perhaps something else is causing this issue?
I think I figured out what the problem is. The RingBuffer was not properly initializing its data members. This way the command flood protection misbehaved. I have a fix and will commit this soon.

TorrSamaho commented:
I added the updated `RingBuffer` and also your patched related to `GAMEEVENT_CHAT`. Still need to think about the cursor code.

AKMDM commented:
All commits including the blinking chat cursor have been added in, so this PR can be closed now.
